added new metric: contribution-distance

JustinGOSSES opened the pull request:
Added new metric: contribution-distance

Attempted to add it to the mermaid.js diagram as well but couldn't get the mermaid diagram to generate locally with or without changes, so that might need a review by @jeffabailey

jeffabailey commented:
@JustinGOSSES I ran the script locally, and it works fine.

You don't need to update use_gqm.md though.

Here's what the graph will look like after GitHub actions generates the diagram and commits the change after the PR is merged.

Basically, adding new goals, questions, and metrics while linking the to each other is all you have to do as a contributor.

<img width="778" alt="image" src="https://github.com/InnerSourceCommons/managing-innersource-projects/assets/776901/9433f2a5-ceef-4488-a8c3-232d7c4c2923">

JustinGOSSES commented:
I wanted there to be an arrow from the contribution-distance block to the question block about understanding who contributes to a project, but it looks like that arrow didn't appear?

I'll try to new generation instructions and report back.

jeffabailey commented:
> I wanted there to be an arrow from the contribution-distance block to the question block about understanding who contributes to a project, but it looks like that arrow didn't appear? There is a line in the mermaid diagram code for the arrow but I'll take another look.
> 
> I'll try to new generation instructions and report back.

If you add another link to the bottom of this file and rename the metric's file to contribution-distance.md it should generate the link as expected.

https://github.com/InnerSourceCommons/managing-innersource-projects/blob/main/measuring/questions/who-uses.md?plain=1#L15

The gqm_gen script expects hyphenated filenames rather than underscore delimited.

JustinGOSSES commented:
Addressed all the issues, generated diagram successfully locally, and added to the instructions for generating the diagram locally to address some problems I ran into: 
- Need to install mermaid.js globally as it doesn't get installed by the package.json
- The .mmd instead of .md file ending confusing me as I didn't expect it, so mentioned that in instructions. 
- Added some language to explicitly describe all the files that need to be possibly edited when making changes to the diagram. 
- Added a warning about not using `_` underscores in filenames and using `-` hyphens instead.

JustinGOSSES commented:
![image](https://github.com/InnerSourceCommons/managing-innersource-projects/assets/11600445/5c711b44-486b-4c2e-91db-e323ac564e48)

jeffabailey commented on scripts/gqm_gen/README.md:
````diff
@@ -2,11 +2,53 @@
 ```bash
-npm install
-npm run --silent start
+npm run --silent start > gqm.mmd && mmdc --input gqm.md --output gqm.png
````
Did you need to create an .mmd file to preview with your editor?

In vscode, it's possible to preview the .md file.

<img width="1068" alt="image" src="https://github.com/InnerSourceCommons/managing-innersource-projects/assets/776901/6222a019-a2a2-4261-9b50-d9baddef0f5c">


```suggestion 
npm run --silent start > gqm.md && mmdc --input gqm.md --output gqm.png
```

jeffabailey commented on scripts/gqm_gen/README.md:
```diff
@@ -2,11 +2,53 @@
+
+This should generate a file called `gqm.png` or `gqm-1.png` in the same folder as this file.
+
+You will also end up with BOTH a `gqm.md` and a `gqm.mmd` file as the later is created upon running the script.
```
No mmd file should be generated after adjusting the recommended command above.

<img width="936" alt="image" src="https://github.com/InnerSourceCommons/managing-innersource-projects/assets/776901/1c4a5d34-05ee-463a-8266-ff565d95c025">

JustinGOSSES replied:
changed as suggested

jeffabailey commented on scripts/gqm_gen/gqm-1.png:
Need to delete this file, and we may want to add this to the bottom of the .gitignore file.

innersource-commons/managing-innersource-projects/scripts/gqm_gen/.gitignore

```
gqm*[pngmd]
```

JustinGOSSES replied:
Deleted

rrrutledge commented:
Hey @JustinGOSSES, sorry I'm behind on my _GitHub_ review queue so just looking at this.  It looks like you and @jeffabailey have had a bit of back-and-forth, so just wanted to ping you to see if you can have a look at Jeff's requested changes?  Excited to get this in!

JustinGOSSES commented on scripts/gqm_gen/gqm.md:
````diff
@@ -0,0 +1,37 @@
+```mermaid
````
done

JustinGOSSES commented on scripts/gqm_gen/gqm.mmd:
````diff
@@ -0,0 +1,37 @@
+```mermaid
````
Done
